### PR TITLE
No did-you-mean-to-call error on casts

### LINKED ISSUE #41640: [NewErrors] 4.2.0-dev.20201122 vs 4.1.2
The following errors were reported by 4.2.0-dev.20201122, but not by 4.1.2

# [ant-design](https://github.com/ant-design/ant-design)
### [tsconfig.json](https://github.com/ant-design/ant-design/blob/87da555ef096489972655042ff765b893020bcb7/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [components/affix/index.tsx#L240](https://github.com/ant-design/ant-design/blob/87da555ef096489972655042ff765b893020bcb7/components/affix/index.tsx#L240)
   - [components/layout/Sider.tsx#L115](https://github.com/ant-design/ant-design/blob/87da555ef096489972655042ff765b893020bcb7/components/layout/Sider.tsx#L115)
   - [components/upload/Upload.tsx#L200](https://github.com/ant-design/ant-design/blob/87da555ef096489972655042ff765b893020bcb7/components/upload/Upload.tsx#L200)
   - [site/theme/template/IconDisplay/IconPicSearcher.tsx#L112](https://github.com/ant-design/ant-design/blob/87da555ef096489972655042ff765b893020bcb7/site/theme/template/IconDisplay/IconPicSearcher.tsx#L112)
# [better-scroll](https://github.com/ustbhuangyi/better-scroll)
### [test-dts/tsconfig.json](https://github.com/ustbhuangyi/better-scroll/blob/a482ad7acf2bc8ae6c436307a5b6f2ca1a2a58a5/test-dts/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - file:///mnt/ts_downloads/better-scroll/node_modules/@better-scroll/shared-utils/src/lang.ts#L3
   - [packages/shared-utils/src/lang.ts#L3](https://github.com/ustbhuangyi/better-scroll/blob/a482ad7acf2bc8ae6c436307a5b6f2ca1a2a58a5/packages/shared-utils/src/lang.ts#L3)
### [tsconfig.json](https://github.com/ustbhuangyi/better-scroll/blob/a482ad7acf2bc8ae6c436307a5b6f2ca1a2a58a5/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/shared-utils/src/lang.ts#L3](https://github.com/ustbhuangyi/better-scroll/blob/a482ad7acf2bc8ae6c436307a5b6f2ca1a2a58a5/packages/shared-utils/src/lang.ts#L3)
# [turbolinks](https://github.com/turbolinks/turbolinks)
### [tsconfig.json](https://github.com/turbolinks/turbolinks/blob/71b7a7d0546a573735af99113b622180e8a813c2/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [src/controller.ts#L21](https://github.com/turbolinks/turbolinks/blob/71b7a7d0546a573735af99113b622180e8a813c2/src/controller.ts#L21)
   - [src/controller.ts#L22](https://github.com/turbolinks/turbolinks/blob/71b7a7d0546a573735af99113b622180e8a813c2/src/controller.ts#L22)
# [redux-devtools](https://github.com/reduxjs/redux-devtools)
**16 of 50 projects failed to build with the old tsc**
### [packages/map2tree/test/tsconfig.json](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/map2tree/test/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/map2tree/src/index.ts#L66](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/map2tree/src/index.ts#L66)
### [packages/map2tree/tsconfig.json](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/map2tree/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/map2tree/src/index.ts#L66](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/map2tree/src/index.ts#L66)
### [packages/redux-devtools-instrument/tsconfig.json](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-instrument/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/redux-devtools-instrument/src/instrument.ts#L164](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-instrument/src/instrument.ts#L164)
### [packages/redux-devtools-core/test/tsconfig.json](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-core/test/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/redux-devtools-core/src/utils/filters.ts#L146](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-core/src/utils/filters.ts#L146)
### [packages/redux-devtools-core/tsconfig.json](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-core/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [packages/redux-devtools-core/src/utils/filters.ts#L146](https://github.com/reduxjs/redux-devtools/blob/c980eadec70457f3713990954f43c3a3e36a6dd3/packages/redux-devtools-core/src/utils/filters.ts#L146)
# [brain.js](https://github.com/BrainJS/brain.js)
**1 of 2 projects failed to build with the old tsc**
### [tsconfig.json](https://github.com/BrainJS/brain.js/blob/226590805bff7f841e92892edd7006edc5195f87/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [src/train-stream.ts#L132](https://github.com/BrainJS/brain.js/blob/226590805bff7f841e92892edd7006edc5195f87/src/train-stream.ts#L132)
   - [src/train-stream.ts#L135](https://github.com/BrainJS/brain.js/blob/226590805bff7f841e92892edd7006edc5195f87/src/train-stream.ts#L135)
# [redoc](https://github.com/Redocly/redoc)
### [tsconfig.json](https://github.com/Redocly/redoc/blob/c6b7d567d97201d2c244915a035f8c30b1152e4e/tsconfig.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [src/services/ClipboardService.ts#L18](https://github.com/Redocly/redoc/blob/c6b7d567d97201d2c244915a035f8c30b1152e4e/src/services/ClipboardService.ts#L18)
### [tsconfig.lib.json](https://github.com/Redocly/redoc/blob/c6b7d567d97201d2c244915a035f8c30b1152e4e/tsconfig.lib.json)
 - `error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?`
   - [src/services/ClipboardService.ts#L18](https://github.com/Redocly/redoc/blob/c6b7d567d97201d2c244915a035f8c30b1152e4e/src/services/ClipboardService.ts#L18)

### PR DESCRIPTION
**Pull request by sandersn**
I chose to do the ad-hoc check rather than yet another tree walk.

1. It's faster to run and easier to read.
2. This error came from looking at real code. It happened only twice, so I think the best estimate for other ways people have included a cast is zero, given that they've happened zero times so far.
3. I couldn't think of other places to put the cast, given the restrictions on `testedNode` just before the new code.

Fixes #41640

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -34554,8 +34554,11 @@ namespace ts {
                 : isPropertyAccessExpression(location) ? location.name
                 : isBinaryExpression(location) && isIdentifier(location.right) ? location.right
                 : undefined;
+            const isPropertyExpressionCast = isPropertyAccessExpression(location)
+                && isParenthesizedExpression(location.expression)
+                && isAssertionExpression(location.expression.expression);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:34559]: It is weird that adding a layer of parens will throw this off. Maybe use `skipParentheses`

**Comment by DanielRosenwasser** [comment]
Will this need to be cherry-picked into 4.2 stable?

**Review by elibarzilay — APPROVED** [review]
Related to @DanielRosenwasser's comment, I noticed that this:

```ts
function f(result: unknown) {
    if (((result as I).always)) {
        return result
    }
}
```

doesn't throw the error.  (So if it's just a rare case, it could be used
to avoid it?)

(Just a comment, I don't have an opinion on these errors...)

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts:34559]: Good idea

**Comment by sandersn** [comment]
@elibarzilay Casts are a usual way for me to try to silence an error I know is bogus, so I think it's important that this works. 
@DanielRosenwasser I'd have to read up on the history to decide whether it's important enough to put in 4.2.

**Comment by sandersn** [comment]
OOPS. I merged this without the followup commit. I opened a new PR here: https://github.com/microsoft/TypeScript/pull/42779

**Comment by DanielRosenwasser** [comment]
I think it's probably safe to error *less*. Pinging @RyanCavanaugh for thoughts.

In the meantime...

@typescript-bot cherry-pick this to release-4.2

**Pull request merged**